src/layer/marker/Icon.Default.js

```diff
@@ -19,42 +19,133 @@ import * as DomUtil from '../../dom/DomUtil';
+		// the blue icon images. If you are placing these images in a
+		// non-standard way, set this option to point to the right
+		// path, before any marker is added to a map.
+		// Caution: do not use this option with inline base64 image(s).
+		var imagePath = this.options.imagePath || IconDefault.imagePath || '';
+		// Deprecated (IconDefault.imagePath), backwards-compatibility only
+
+		if (this._needsInit) {
+			// Modifying imagePath option after _getIconUrl has been called
+			// once in this instance of IconDefault will no longer have any
+			// effect.
+			this._initializeOptions(imagePath);
 		}
 
-		// @option imagePath: String
-		// `Icon.Default` will try to auto-detect the location of the
-		// blue icon images. If you are placing these images in a non-standard
-		// way, set this option to point to the right path.
-		return (this.options.imagePath || IconDefault.imagePath) + Icon.prototype._getIconUrl.call(this, name);
+		return imagePath + Icon.prototype._getIconUrl.call(this, name);
 	},
 
-	_detectIconPath: function () {
-		var el = DomUtil.create('div',  'leaflet-default-icon-path', document.body);
-		var path = DomUtil.getStyle(el, 'background-image') ||
-		           DomUtil.getStyle(el, 'backgroundImage');	// IE8
+	// Initialize all necessary options for this instance.
+	_initializeOptions: function (imagePath) {
+		this._setOptions('icon', _detectIconOptions, imagePath);
+		this._setOptions('shadow', _detectIconOptions, imagePath);
+		this._setOptions('popup', _detectDivOverlayOptions);
+		this._setOptions('tooltip', _detectDivOverlayOptions);
+		this._needsInit = false;
+	},
 
-		document.body.removeChild(el);
+	// Retrieve values from CSS and assign to this instance options.
+	_setOptions: function (name, detectorFn, imagePath) {
+		var options = this.options,
+		    prefix = options.classNamePrefix,
+		    optionValues = detectorFn(prefix + name, imagePath);
 
-		if (path === null || path.indexOf('url') !== 0) {
-			path = '';
-		} else {
-			path = path.replace(/^url\(["']?/, '').replace(/marker-icon\.png["']?\)$/, '');
+		for (var optionName in optionValues) {
+			options[name + optionName] = options[name + optionName] || optionValues[optionName];
 		}
-
-		return path;
 	}
 });
+
+// Retrieve icon option values from CSS (icon or shadow).
+function _detectIconOptions(className, imagePath) {
+	var el = DomUtil.create('div',  className, document.body),
+	    urlsContainer = _getBkgImageOrCursor(el),
+	    urls = _extractUrls(urlsContainer, imagePath),
+	    iconX = _getStyleInt(el, 'width'),
+	    iconY = _getStyleInt(el, 'height'),
+	    anchorNX = _getStyleInt(el, 'margin-left'),
+	    anchorNY = _getStyleInt(el, 'margin-top');
+
+	return {
+		Url: urls[0],
+		RetinaUrl: urls[1],
+		Size: [iconX, iconY],
+		Anchor: [-anchorNX, -anchorNY]
+	};
+}
+
+// Retrieve anchor option values from CSS (popup or tooltip).
+function _detectDivOverlayOptions(className) {
+	var el = DomUtil.create('div', className, document.body),
```

Comment:
Remove 3rd argument `document.body`, otherwise the dummy element will actually be added to the document…